c++/error-handling.md

```diff
@@ -0,0 +1,672 @@
+which violates the zero-overhead principle in the case of a bug-free program.
+
+Broadly, there are three possible strategies that can be adopted
+at the point where an assertion would evaluate to `false`:
+
+### Log-And-Continue Enforcement Strategy
+
+The assert statement could emit a run-time diagnostic regarding the contract violation
+and then continue past the assertion. This is not ideal: the program is known to
+exhibit undefined behaviour at this point.
+
+There is a strong desire to distinguish between contract violations that
+exhibit UB and ones which are somehow safe. But this distinction is false.
+An assert expression needs to be something that is inconceivable. As such, its cause
+can only be explained by some unforeseen error earlier in the control flow.
+This in turn leaves the door open to UB caused by the error elsewhere.
+The horse has already bolted, so to speak.
+
+Nevertheless, under certain constraints this strategy is appealing. If the cost of
+terminating the program is greater than the risks of undefined behaviour, then
+a programmer might choose this path. It would be sensible to do everything possible
+to mitigate the effects, e.g. by disabling optimisations.
+
+### Trap Enforcement Strategy
+
+The assert statement can halt the program.
+
+This has the disadvantage that the program completely fails to perform its task.
+But the advantage is that no further incorrect behaviour
+will be exhibited by the program.
+
+### Prevention Enforcement Strategy
+
+Assuming bugs are eliminated, no assert expression will ever be false.
+This means that assert statements behave as if they are not there.
+This in turn means that they can be empty or used as hints to the optimiser.
+
+There is concern ([P2064](https://wg21.link/p2064r0)) that assuming no violations
+is somehow incorrect. This is founded on the acceptance of bugs in production code
+and the assumption that optimisers *cause* undefined behaviour. They do not. Bugs
+cause undefined behaviour. If developers wish to release code with bugs to production,
+they are welcome to disable such assumptions. But they are invited, instead, to follow
+practices described as part of the Test User Contract which should help identify
+such bugs early on.
+
+***The author strongly encourages the use of assumptions in the implementation of
+assertions in order to unify the set of Dynamically-Enforceable Contracts and to
+simplify the implementation of the Test User Contract.***
+In practice, this entails marking violated assertions as unreachable
+
+```c++
+#define ASSERT(cond) ((cond) ? static_cast<void>(0) : __builtin_unreachable())
+```
+
+which is semantically sound, given violations already must never occur in correctly-written
+C++ programs.
+
+This strategy provides the maximum leeway to support enforcement profiles (below)
+and ensures that the existing ability of sanitizers is extended,
+and that constant expressions containing bugs become ill-formed.
+
+## Enforcement Profiles
```

Comment:
This phrase doesn't convey what it is supposed to and can be confused with the strategies. Rename. Suggestion: Dynamic Contract Enforcement Profiles.